README
A GUI for Missing Data Exploration

Install
------------

GTK+ is required to install this package.
To install GTK+, try the following line in R:

```
install.packages('RGtk2')
```

If any error occurs, then run the above line again.
If GTK+ can still not be installed, visit http://www.gtk.org/download/ and install GTK+ 2.x.
(For Mac OS users, visit http://r.research.att.com/ to get the recent GTK+ 2.x framework binary.)
Then run the above R code again.

After installing the RGtk2 package, run

```
install.packages('MissingDataGUI')
```



NEWS
Changes in Version 0.2-5

  - Fix the color bugs in pairwise plot and parallel coordinates plot
  

Changes in Version 0.2-4

  - Small changes based on the new version of package mi
  - Add citation
  - Update the affiliation information 


Changes in Version 0.2-3

  - Small changes based on the new version of package mi


Changes in Version 0.2-2

  - In the 'Simple' method, 'Mean/Mode' is renamed by 'Mean'. Mode applies when median or mean does not work.
  - Update the parameter for hclust (from 'ward' to 'ward.D')
  - mi from the mi package only accepts n.imp>1, so we set n.imp=2 when the actual input is 1, and take the first set as the result.


Changes in Version 0.2-0

NEW FEATURES

  - Add new imputation methods (multiple imputations and nearest neighbors), re-arrange the widget for methods.

  - Involve the tabs for the graphic device, which can display different methods or multiple chains in parallel.

  - Add a new graph for parallel coordinates plot, which sorts the variables by F-statistics of ANOVA.

  - Add a "Setting" tab to tune the parameters.

  - Fix the bugs in the graphical device for Windows platform.


Changes in Version 0.1-6

Edit NAMESPACE by required.


Changes in Version 0.1-4

NEW FEATURES

  - A new window is created when hitting the button "Export Data" on the main GUI. This window provides the following options for exporting the imputed data:

    * Whether to export the data of all the variables or only the selected variables. The default setting is all variables. If the user does not select anything when pressing the "Export Data" button, then all the variables are highlighted/selected.

    * Whether to export the shadow matrix. Shadow matrix is useful to indicate the places of missing values / imputed values after the imputed data replace NA's.

    * Where to save the exported data, i.e. to choose the directory.

    * What name to use for the exported data.

    * What format to save the data. Users can save them in a csv format or rda format file, or export the data to R console directly as a data frame.


R/utils.r
if(getRversion() >= '2.15.1') globalVariables(c("observation", "variable","Reordered_Observation","Missing","x1","x2","y1","y2"))

## The default color palette for MissingDataGUI
##
## Service the colorblind.
## \url{http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/}
## @name cbPalette
## @docType data
## @noRd
##
#cbPalette = c("#56B4E9", "#E69F00", "#009E73", "#000000")
##' Change the discrete color scale for the plots generated by ggplot2
##'
##' @param ... parameters passed into the function
##' @return NULL
##' @export
##'
#scale_colour_discrete = function(...) scale_colour_manual(values=cbPalette)
scale_colour_discrete = function(...) scale_colour_manual(values=c("#56B4E9", "#E69F00"))
##' Change the discrete fill scale for the plots generated by ggplot2
##'
##' @param ... parameters passed into the function
##' @return NULL
##' @export
##'
#scale_fill_discrete = function(...) scale_fill_manual(values=cbPalette)
scale_fill_discrete = function(...) scale_fill_manual(values=c("#56B4E9", "#E69F00"))
# ColorBrewer2.org
# scale_colour_discrete = function(...) scale_colour_brewer(..., type="qual",palette="Dark2")
# scale_fill_discrete = function(...) scale_fill_brewer(..., type="qual",palette="Dark2")


.onLoad = function(lib, pkg) {
  options(guiToolkit = "RGtk2")
  library_call('gWidgetsRGtk2')
}

##' Load the package when necessary
##'
##' @param pkg a character of the package name.
##' @return NULL
##' @author Xiaoyue Cheng <\email{xycheng@@unomaha.edu}>
##' @noRd
##'
library_call = function(pkg) {
  library(pkg, character.only = TRUE)
}

##' Install the package when necessary
##'
##' @param pkg a character of the package name.
##' @return NULL
##' @author Xiaoyue Cheng <\email{xycheng@@unomaha.edu}>
##' @noRd
##'
library_gui = function(pkg) {
  if (require(pkg, character.only=TRUE)) return()
  if (gconfirm(paste('Install the missing package ', pkg, '?', sep = ''))) {
    install.packages(pkg)
    library_call(pkg)
  } else stop('The package ', pkg, ' is not available')
}

##' Compute the numeric summary of the missingness
##'
##' @param dat A data frame.
##' @return A list including (1) a data frame 'missingsummary' that provides
##' a table of missingness; (2) the total missing percentage; (3) the percent
##' of variables that contain missing values; (4) the ratio of observations
##' that have missings.
##' @author Xiaoyue Cheng <\email{xycheng@@unomaha.edu}>
##' @export
##' @examples
##' data(tao)
##' compute_missing_pct(tao)
##'
compute_missing_pct = function(dat){
    stopifnot(is.data.frame(dat))
    n = ncol(dat)
    totalmissingpct = mean(is.na(dat))
    varmissingpct = mean(sapply(dat,function(avec){any(is.na(avec))}))
    casemissingpct = 1-mean(complete.cases(dat))
    No_of_Case_missing = table(apply(dat, 1, function(avec){sum(is.na(avec))}))
    No_of_Case = rep(0,(n+1))
    No_of_Case[n+1-as.integer(names(No_of_Case_missing))]=No_of_Case_missing[names(No_of_Case_missing)]
    No_of_Case[n+1] = sum(complete.cases(dat))
    missingsummary = data.frame(No_of_miss_by_case=n:0,
                                No_of_Case=No_of_Case,
                                Percent=as.character(round(No_of_Case/nrow(dat)*100,1)))
    missingsummary = missingsummary[order(missingsummary$No_of_miss_by_case, decreasing=FALSE),]
    return(list(missingsummary=missingsummary,totalmissingpct=totalmissingpct,
                varmissingpct=varmissingpct,casemissingpct=casemissingpct))
}

##' Find the default methods given by \pkg{mice}
##'
##' @param vec a vector of the variable classes.
##' @param dat the data corresponding to vec.
##' @return A vector of the default imputation method generated by mice.
##' @author Xiaoyue Cheng <\email{xycheng@@unomaha.edu}>
##' @noRd
##'
mice_default = function(vec, dat){
    res = vec
    res[res %in% c("integer","numeric")] = "pmm"
    res[res == "logical"] = "logreg"
    for (i in which(vec %in% c('factor','character','ordered'))) {
        s = length(unique(na.omit(dat[,i])))
        if (s<3) {res[i] = "logreg"} else {
            res[i] = if (vec[i] == "ordered") {"polr"} else {"polyreg"}
        }
    }
    res[colSums(is.na(dat))==0] = ""
    return(res)
}

##' Print ggpair object in GTK+ for Windows platform
##'
##' See \code{print.ggpairs} from \pkg{GGally}.
##'
##' @param x ggpair object to be plotted
##' @param ... not used
##' @return NULL
##' @author Barret Schloerke <\email{schloerke@@gmail.com}>, Xiaoyue Cheng <\email{xycheng@@unomaha.edu}>
##' @importFrom GGally getPlot
##' @importFrom grid viewport unit grid.layout grid.newpage pushViewport grid.text gpar popViewport grid.rect
##' @noRd
##'
winprint = function (x, ...) {
  plotObj <- x
  if (identical(plotObj$axisLabels, "internal")) {
    v1 <- viewport(y = unit(0.5, "npc") - unit(0.5, "lines"),
                   width = unit(1, "npc") - unit(1, "lines"),
                   height = unit(1, "npc") - unit(2, "lines"))
  }
  else {
    v1 <- viewport(width = unit(1, "npc") - unit(3, "lines"),
                   height = unit(1, "npc") - unit(3, "lines"))
  }
  numCol <- length(plotObj$columns)
  v2 <- viewport(layout = grid.layout(numCol, numCol, widths = rep(1, numCol), heights = rep(1, numCol)))
  grid.newpage()
  if (plotObj$title != "") {
    pushViewport(viewport(height = unit(1, "npc") - unit(0.4, "lines")))
    grid.text(plotObj$title, x = 0.5, y = 1, just = c(0.5, 1), gp = gpar(fontsize = 15))
    popViewport()
  }
  if (!identical(plotObj$axisLabels, "internal")) {
    pushViewport(viewport(width = unit(1, "npc") - unit(2, "lines"),
                          height = unit(1, "npc") - unit(3, "lines")))
    pushViewport(viewport(layout = grid.layout(numCol, numCol, widths = rep(1, numCol),
                                               heights = rep(1, numCol))))
    for (i in 1:numCol) {
      grid.text(names(plotObj$data[, plotObj$columns])[i],
                0, 0.5, rot = 90, just = c("centre", "centre"),
                vp = viewport(layout.pos.row = as.numeric(i), layout.pos.col = 1))
    }
    popViewport()
    popViewport()
    pushViewport(viewport(width = unit(1, "npc") - unit(3, "lines"),
                          height = unit(1, "npc") - unit(2, "lines")))
    pushViewport(viewport(layout = grid.layout(numCol, numCol, widths = rep(1, numCol),
                                               heights = rep(1, numCol))))
    for (i in 1:numCol) {
      grid.text(names(plotObj$data[, plotObj$columns])[i],
                0.5, 0, just = c("centre", "centre"),
                vp = viewport(layout.pos.row = numCol, layout.pos.col = i))
    }
    popViewport()
    popViewport()
  }
  pushViewport(v1)
  pushViewport(v2)
  for (rowPos in 1:numCol) {
    for (columnPos in 1:numCol) {
      p <- getPlot(plotObj, rowPos, columnPos)
      p_cond <- if( !is.null(p$subType) && !is.null(p$type)){
        p$subType == "blank" && p$type == "blank"
      } else {FALSE}
      if (!p_cond) {
        pos <- columnPos + (rowPos - 1) * numCol
        type <- p$type
        subType <- p$subType
        if (plotObj$printInfo) {
          cat("Pos #", pos)
          if (!is.null(type))
            cat(": type = ", type)
          if (!is.null(subType))
            cat(": subType = ", subType)
          cat("\n")
        }
        noTicks <- c("internal", "none")
        removeTicks <- plotObj$axisLabels %in% noTicks
        if (!is.null(p$axisLabels)) {
          removeTicks <- p$axisLabels %in% noTicks
        }
        if (columnPos != 1 || removeTicks) {
          p <- p + theme(axis.text.y = element_blank(),
                         axis.title.y = element_blank())
        }
        if ((rowPos != numCol) || removeTicks) {
          p <- p + theme(#axis.text.x = element_blank(),
            axis.text.x = element_text(colour = "white"),
            axis.title.x = element_blank())
        }
        if (removeTicks) {
          p <- p + theme(strip.background = element_rect(fill = "white", colour = NA),
                         strip.text.x = element_blank(), strip.text.y = element_blank(),
                         axis.ticks = element_blank())
        }
        if (identical(p$type, "combo")) {
          p <- p + labs(x = NULL, y = NULL)
          if (plotObj$printInfo) {
            print(p$subType)
            print(p$horizontal)
          }
          if (p$horizontal) {
            if (p$subType %in% c("facethist")) {
              p <- p + theme(plot.margin = unit(c(0, -0.5, 0, 0), "lines"))
            }
            else {
              p <- p + theme(plot.margin = unit(c(0, -0.5, 0, -0.5), "lines"))
            }
            if (columnPos != numCol) {
              p <- p + theme(strip.background = element_blank(),
                             strip.text.x = element_blank(), strip.text.y = element_blank())
            }
          }
          else {
            if (p$subType %in% c("facethist")) {
              p <- p + theme(plot.margin = unit(c(-0.5, 0, 0, 0), "lines"))
            }
            else {
              p <- p + theme(plot.margin = unit(c(-0.5, 0, -0.5, 0), "lines"))
            }
            if (rowPos != 1) {
              p <- p + theme(strip.background = element_blank(),
                             strip.text.x = element_blank(), strip.text.y = element_blank())
            }
          }
        }
        else if (identical(p$subType, "facetbar")) {
          p <- p + labs(x = NULL, y = NULL) + theme(plot.margin = unit(c(0, -0.5, 0, 0), "lines"))
          if (rowPos != 1) {
            p <- p + theme(strip.background = element_blank(),
                           strip.text.x = element_blank(), strip.text.y = element_blank())
          }
        }
        else if (identical(p$type, "continuous") && !identical(p$subType, "cor")) {
          p <- p + labs(x = NULL, y = NULL) + theme(plot.margin = unit(rep(0, 4), "lines"))
        }
        else if (identical(p$type, "diag") && is.numeric(p$data[, as.character(p$mapping$x)])) {
          p <- p + labs(x = NULL, y = NULL) + theme(plot.margin = unit(rep(0, 4), "lines"))
        }
        else {
          p <- p + labs(x = NULL, y = NULL) + theme(plot.margin = unit(rep(0, 4), "lines"))
        }
        showLegend = FALSE
        if (!is.null(plotObj$legends))
          showLegend <- identical(plotObj$legends, TRUE)
        if (showLegend == FALSE) {
          if (!is.null(p$ggally$legend) && !is.na(p$ggally$legend)) {
            showLegend <- identical(p$ggally$legend, TRUE)
          }
        }
        if (showLegend == FALSE) {
          p <- p + theme(legend.position = "none")
        }
        grid.rect(gp = gpar(fill = "white", lty = "blank"),
                  vp = viewport(layout.pos.row = rowPos, layout.pos.col = columnPos))
        if (identical(plotObj$verbose, TRUE)) {
          print(p, vp = viewport(layout.pos.row = rowPos, layout.pos.col = columnPos))
        }
        else {
          suppressMessages(suppressWarnings(print(p, vp = viewport(layout.pos.row = rowPos, layout.pos.col = columnPos))))
        }
      }
    }
  }
  popViewport()
  popViewport()
}


R/imputation.r
##' Impute the missing data with the method selected under the
##' condition.
##'
##' This function provides eight methods for imputation with
##' categorical varaibles as conditions.
##'
##' The imputation methods: This list displays all the imputation
##' methods. Users can make one selection. (1) 'Below 10%' means
##' NA's of one variable will be replaced by the value which
##' equals to the minimum of the variable minus 10% of the range.
##' For categorical variables, NA's are treated as a new category.
##' Under this status the selected conditioning variables are
##' ignored. If the data are already imputed, then this item
##' will show the imputed result. (2) 'Simple' will create three
##' tabs: Median, Mean, and Random Value. 'Median' means NA's will be
##' replaced by the median of this variable (omit NA's).
##' 'Mean' means NA's will be replaced by the mean of the
##' variable (omit NA's). The median does not apply to the nominal
##' variable, neither does the mean to the categorical variable.
##' In these cases the mode (omit NA's) is provided.
##' 'Random Value' means NA's will be replaced by any values
##' of this variable (omit NA's) which are randomly selected.
##' (3) 'Neighbor' contains two methods: 'Average Neighbor' and
##' 'Random Neighbor'. 'Average Neighbor' will replace the
##' NA's by the mean of the nearest neighbors. 'Random Neighbor'
##' substitutes the missing for a random sample of the k nearest
##' neighbors. The number of neighbors is default to 5, and
##' can be changed by argument knn. The Neighbor methods
##' require at lease one case to be complete, at least two
##' variables to be selected, and no factor/character variables.
##' The ordered factors are treated as integers. The method will
##' return the overall mean or a global random sample value
##' if the observation only contains NA's.
##' (4) 'MI:areg' uses function \code{\link[Hmisc]{aregImpute}}
##' from package \pkg{Hmisc}. It requires at lease one case to
##' be complete, and at least two variables to be selected.
##' (5) 'MI:norm' uses function \code{\link[norm]{imp.norm}}
##' from package \pkg{norm}. It requires all selected variables
##' to be numeric(at least integer), and at least two variables
##' to be selected. Sometimes it cannot converge, then the
##' programme will leave NA's without imputation.
##' (6) 'MI:mice' uses the \pkg{mice} package. The methods of
##' the variables containing NA's must be attached with argument
##' method. If not, then default methods are used.
##' (7) 'MI:mi' employes the \pkg{mi} package.
##' @param origdata A data frame whose missing values need to be
##' imputed. This data frame should be selected from the missing
##' data GUI.
##' @param method The imputation method selected from the missing
##' data GUI. Must be one of 'Below 10%','Simple','Neighbor',
##' 'MI:areg','MI:norm','MI:mice','MI:mi'. If method='MI:mice',
##' then the methods of the variables containing NA's must be attached
##' with argument method. If not, then default methods are used.
##' @param vartype A vector of the classes of origdata. The length is
##' the same as the number of columns of origdata. The value should be
##' from "integer", "numeric", "logical", "character", "factor", and "ordered".
##' @param missingpct A vector of the percentage of missings of the
##' variables in origdata. The length is the same as the number of
##' columns of origdata. The values should be between 0 and 1.
##' @param condition A vector of categorical variables. The dataset
##' will be partitioned based on those variables, and then the
##' imputation is implemented in each group. There are no missing
##' values in those variables. If it is null, then there is no
##' division. The imputation is based on the whole dataset.
##' @param knn number of the neighbors.
##' @param mi.n number of the imputation sets for multiple imputation
##' @param mi.seed random number seed for multiple imputation
##' @param row_var A column name (character) that defines the ID of rows.
##' @return The imputed data frame with the last column being the row
##' number from the original dataset. During the procedure of the
##' function, rows may be exchanged, thus a column of row number could
##' keep track of the original row number and then help to find the
##' shadow matrix.
##' @export
##' @author Xiaoyue Cheng <\email{xycheng@@iastate.edu}>
imputation = function(origdata, method, vartype=NULL, missingpct=NULL, condition=NULL, knn=5, mi.n=3, mi.seed=1234567, row_var=NULL){
    if (is.null(origdata)) return(NULL)
    if (is.null(vartype)) vartype=unname(sapply(origdata,function(x)class(x)[1]))
    if (is.null(missingpct)) missingpct=unname(sapply(origdata,function(x)mean(is.na(x))))
    row_NO = if (is.null(row_var)) {1:nrow(origdata)} else {origdata[,row_var]}
    origdata = origdata[,setdiff(colnames(origdata),row_var),drop=FALSE]
    n=ncol(origdata)
    if (n==1 && substr(method,1,3) == 'MI:'){
        gmessage('You only selected one variable. Cannot apply the multiple imputation.', icon = "error")
        return(NULL)
    }
    if (!is.null(condition)) {
      if (sum(is.na(condition))) {
        condition = sapply(condition,as.character)
        condition[is.na(condition)] = "NaN"
        condition = data.frame(condition)
      }
        dat = by(data.frame(origdata,row_id=1:nrow(origdata)),condition,imputation,method=method,vartype=vartype,missingpct=NULL,condition=NULL,knn=knn,mi.n=mi.n,mi.seed=mi.seed,row_var='row_id',simplify=FALSE)
        dat = dat[!sapply(dat,is.null)]
        k1=length(dat[[1]])
        k2=length(dat)
        res=list()
        for (j in 1:k1) res[[j]]=lapply(dat,function(x) x[[j]])
        names(res)=names(dat[[1]])
        res=lapply(res, function(x){
            tmp=NULL
            for (i in 1:k2) {
                tmp=rbind(tmp,x[[i]])
            }
            return(tmp)
        })
        res = lapply(res,function(x)x[order(x$row_number),c(colnames(origdata),'row_number')])
        return(res)
    }

    dat = list(d1=origdata)
    origshadow = is.na(origdata[,1:n,drop=FALSE])
    if (method == 'Below 10%') {
        for (i in 1:n) {
            if (vartype[i] %in% c('integer','numeric','logical') &
                    as.numeric(as.character(missingpct))[i]>0) {
                dat$d1[origshadow[,i],i] = min(origdata[,i], na.rm=TRUE)*1.1-
                    max(origdata[,i],na.rm=TRUE)/10
            }
            if (vartype[i] == "character") {
                dat$d1[origshadow[,i],i] = 'NAN'
            }
            if (vartype[i] %in% c("factor","ordered") &&
                    as.numeric(as.character(missingpct))[i]>0) {
                dat$d1[,i]=factor(dat$d1[,i],levels = c('NAN',levels(factor(dat$d1[,i]))))
                dat$d1[origshadow[,i],i] = 'NAN'
            }
        }
        names(dat)="Below 10%"
    }
    else if (method == 'Simple') {
      set.seed(mi.seed)
        dat$d3=dat$d2=dat$d1
        for (i in 1:n) {
          if (sum(is.na(dat$d1[,i]))){
            if (vartype[i] %in% c('integer','numeric')) {
              dat$d1[is.na(dat$d1[,i]),i] = median(dat$d1[,i], na.rm=TRUE)
              dat$d2[is.na(dat$d2[,i]),i] = mean(dat$d2[,i], na.rm=TRUE)
            } else if (vartype[i] %in% c('logical','factor','character')){
              biggroup = names(sort(table(na.omit(dat$d2[,i])),decreasing=TRUE))[1]
              dat$d1[origshadow[,i],i] = biggroup
              dat$d2[origshadow[,i],i] = biggroup
            } else {
              grouplevel = levels(dat$d1[,i])
              dat$d1[origshadow[,i],i] = grouplevel[round(median(as.integer(dat$d1[,i]), na.rm=TRUE))]
              dat$d2[origshadow[,i],i] = names(sort(table(na.omit(dat$d2[,i])),decreasing=TRUE))[1]
            }
            fill=sample(dat$d3[!origshadow[,i],i], sum(origshadow[,i]), replace = TRUE)
            dat$d3[origshadow[,i],i] = fill
          }
        }
        names(dat)=c('Median','Mean','Random Value')
    }
    else if (method == 'Neighbor') {
        set.seed(mi.seed)
        if (n==1 || sum(complete.cases(origdata))==0 || any(c("factor","character") %in% vartype)) {
            if (n==1) warning_message='You only selected one variable. Cannot apply the neighbor imputation. Only the random value imputation is given.'
            if (sum(complete.cases(origdata))==0) warning_message="All the observations have missing values. Cannot find the neighbors."
            if (any(c("factor","character") %in% vartype)) warning_message='Cannot impute by the neighbor with one or more factor or character variables.'
            gmessage(warning_message, icon='warning')
            return(NULL)
        } else {
            dat$d2=dat$d1
            CmpltDat = dat$d1[complete.cases(dat$d1),]
            orderedfactor = which(vartype=="ordered")
            if (length(orderedfactor)) {
              for (j in orderedfactor) {
                CmpltDat[,j]=as.integer(CmpltDat[,j])
                dat$d1[,j]=as.integer(dat$d1[,j])
                dat$d2[,j]=as.integer(dat$d2[,j])
              }
            }
            for (i in which(!complete.cases(dat$d1))){
                usecol = which(!is.na(dat$d1[i,]))
                if (length(usecol)>0){
                    NNdat = CmpltDat
                    a = rbind(dat$d1[i,], NNdat)[,usecol,drop=FALSE]
                    b = apply(a,2,scale)
                    NNdat$distance = dist(b)[1:nrow(NNdat)]
                    kNNdat = NNdat[order(NNdat$distance,decreasing=FALSE),][1:min(knn,nrow(NNdat)),]
                    dat$d1[i,-usecol] = colMeans(kNNdat[,1:n][,-usecol,drop=FALSE])
                    dat$d2[i,-usecol] = kNNdat[,1:n][sample(1:nrow(kNNdat),1),-usecol]
                } else {
                    dat$d1[i,] = sapply(origdata, mean, na.rm=TRUE)
                    dat$d2[i,] = sapply(origdata, function(x)sample(na.omit(x),1))
                }
            }
            if (length(orderedfactor)) {
              for (j in orderedfactor){
                dat$d1[,j] = round(dat$d1[,j])
                l = levels(origdata[,j])
                dat$d1[,j] = ordered(l[dat$d1[,j]],levels=l)
                dat$d2[,j] = ordered(l[dat$d2[,j]],levels=l)
              }
            }
            names(dat)=c('Average Neighbor','Random Neighbor')
        }
    }
    else if (method == 'MI:norm') {
        if (any(c('factor','character','ordered') %in% vartype)) {
            gmessage("Not every variable is numeric. Cannot impute missing values under the multivariate normal model.", icon = "error")
            return(NULL)
        } else {
          library_gui('norm')
            s = norm::prelim.norm(as.matrix(origdata))
            thetahat = norm::em.norm(s)
            norm::rngseed(mi.seed)
            theta = norm::da.norm(s,thetahat,steps=10,showits=TRUE)
            for (i in 1:mi.n){
              dat[[i]] = norm::imp.norm(s,theta,as.matrix(origdata))
            }
            names(dat)=paste('norm',1:mi.n)
            if (any(sapply(dat,function(x){any(c(Inf,NaN) %in% x)}))) {
                gmessage("The algorithm doesn't converge. Return the original data with missing values.", icon = "warning")
                dat = list(original=origdata)
            }
        }
    }
    else if (method == 'MI:areg') {
        if (sum(complete.cases(origdata))==0) {
            gmessage('All the observations have missing values. Cannot impute by Hmisc::aregImpute.', icon = "warning")
            return(NULL)
        } else {
          library_gui('Hmisc')
            formula0 = as.formula(paste('~ ',paste(names(origdata),collapse=' + ')))
            set.seed(mi.seed)
            f = Hmisc::aregImpute(formula0, data=origdata, n.impute=mi.n)
            tmpres = f$imputed
            tmpres = tmpres[!sapply(tmpres,is.null)]
            for (i in 1:mi.n) {
              dat[[i]]=origdata
              for (j in 1:length(tmpres)){
                dat[[i]][rownames(tmpres[[j]]),names(tmpres)[j]]=tmpres[[j]][,i]
              }
            }
            names(dat)=paste('areg',1:mi.n)
        }
    }
    else if (method == 'MI:mice') {
      library_gui('mice')
      if (is.null(attr(method,'method'))) attr(method,'method')=vector("character", length = ncol(data))
        f = mice::mice(origdata, method=attr(method,'method'), m=mi.n, printFlag=FALSE, seed=mi.seed)
        tmpres = f$imp
        for (j in 1:mi.n){
          dat[[j]] = origdata
          for (i in which(!sapply(tmpres,is.null))) {
            dat[[j]][rownames(tmpres[[i]]),names(tmpres)[i]]=tmpres[[i]][,j]
          }
        }
        names(dat)=paste('mice',1:mi.n)
    }
    else if (method == 'MI:mi') {
      library_gui('mi')
        f = mi::mi(origdata, n.chains=mi.n, seed=mi.seed)
        dat = mi::complete(f, include_missing = FALSE)
        names(dat)=paste('mi',1:mi.n)
    }
    res = lapply(dat, function(x) data.frame(x, row_number=row_NO))
    return(res)
}


R/numsummary.r
##' Compute the numeric summary of the missingness
##' 
##' @param dat A data frame.
##' @return A list including (1) a data frame 'missingsummary' that provides 
##' a table of missingness; (2) the total missing percentage; (3) the percent
##' of variables that contain missing values; (4) the ratio of observations
##' that have missings.
##' @author Xiaoyue Cheng <\email{xycheng@@iastate.edu}>
##' @export
##' @examples
##' data(tao)
##' compute_missing_pct(tao)
##' 
compute_missing_pct = function(dat){
    stopifnot(is.data.frame(dat))
    n = ncol(dat)
    totalmissingpct = mean(is.na(dat))
    varmissingpct = mean(sapply(dat,function(avec){any(is.na(avec))}))
    casemissingpct = 1-mean(complete.cases(dat))
    No_of_Case_missing = table(apply(dat, 1, function(avec){sum(is.na(avec))}))
    No_of_Case = rep(0,(n+1))
    No_of_Case[n+1-as.integer(names(No_of_Case_missing))]=No_of_Case_missing[names(No_of_Case_missing)]
    No_of_Case[n+1] = sum(complete.cases(dat))
    missingsummary = data.frame(No_of_miss_by_case=n:0,
                                No_of_Case=No_of_Case,
                                Percent=as.character(round(No_of_Case/nrow(dat)*100,1)))
    missingsummary = missingsummary[order(missingsummary$No_of_miss_by_case, decreasing=FALSE),]
    return(list(missingsummary=missingsummary,totalmissingpct=totalmissingpct,
                varmissingpct=varmissingpct,casemissingpct=casemissingpct))
}


mice_default = function(vec, dat){
    res = vec
    res[res %in% c("integer","numeric")] = "pmm"
    res[res == "logical"] = "logreg"
    for (i in which(vec %in% c('factor','character','ordered'))) {
        s = length(unique(na.omit(dat[,i])))
        if (s<3) {res[i] = "logreg"} else {
            res[i] = if (vec[i] == "ordered") {"polr"} else {"polyreg"}
        }
    }
    res[colSums(is.na(dat))==0] = ""
    return(res)
}


R/SingleImputation.r
##' Impute the Data Without Consideration of Conditions.
##'
##' This function is to impute missing data by one of the eight
##' methods without the condition. When the condtions are imported,
##' they are then romoved from the data frame, and then the imputation
##' will be applied without conditions. Details of the eight methods
##' can be found in \code{\link{imputation}}.
##'
##' See \code{\link{imputation}}.
##'
##' @param dat A data frame whose missing values need to be
##' imputed. The last column should be the row number passed by
##' \code{\link{imputation}}.
##' @param method The imputation method selected from the missing data
##' GUI. Must be one of 'Below 10%','Median','Mean','Random
##' value','Regression','Nearest neighbor','Multiple
##' Imputation','Mode'.
##' @param vartype  A vector of the classes of dat. The length is
##' the same as the number of columns of dat. The value should be
##' from "integer", "numeric", "factor", and "character".
##' @param cond A vector of categorical variables. There are no
##' missing values in those variables.
##' @return The imputed data frame with a column of row numbers
##' matching with the last column of dat, to pass back to
##' \code{\link{imputation}}.
##' @author Xiaoyue Cheng <\email{xycheng@@iastate.edu}>
singleimputation = function(dat, method, vartype, cond) {
  row_number_2011 = dat[,ncol(dat)]
  dat=dat[,-ncol(dat)]
  dat=dat[,setdiff(colnames(dat),cond)]
  if (is.vector(dat)) {n=1} else {n = ncol(dat)}
  shadow = is.na(dat)
  if (n!=1) {
    missingpct = sapply(dat, function(avec){mean(is.na(avec))})
    if (method == 'Below 10%') {
      for (i in 1:n) {
        if (vartype[i] %in% c('integer','numeric','logical') &
          missingpct[i]>0) {
          dat[shadow[,i],i] = min(dat[,i], na.rm=TRUE)*1.1-
            max(dat[,i],na.rm=TRUE)/10
        }
        if (vartype[i] == "character") {
          dat[shadow[,i],i] = 'NAN'
        }
        if (vartype[i] == "factor" & missingpct[i]<1 &
          missingpct[i]>0) {
          dat[,i]=factor(dat[,i],levels = c('NAN',levels(factor(dat[,i]))))
          dat[shadow[,i],i] = 'NAN'
        }
        if (vartype[i] == "factor" & missingpct[i]==0) {
          dat[,i]=factor(dat[,i])
        }
      }
    }
    if (method == 'Median') {
      for (i in 1:n) {
        dat[,i] = as.numeric(impute(dat[,i], fun=median))
      }
    }
    if (method == 'Mean') {
      for (i in 1:n) {
        dat[,i] = as.numeric(impute(dat[,i], fun=mean))
      }
    }
    if (method == 'Random value') {
      for (i in 1:n) {
        dat[,i] = as.numeric(impute(dat[,i], fun='random'))
      }
    }
    if (method == 'Multiple Imputation') {
      if (!(any(c('factor','character') %in% vartype))) {
        s = prelim.norm(as.matrix(dat))
        thetahat = em.norm(s)
        rngseed(1234567)
        dat2 = imp.norm(s,thetahat,as.matrix(dat))
        if (any(sapply(dat2,function(avec){any(c(Inf,NaN) %in% avec)}))) {
          gmessage("This method doesn't converge.
					Leave the NA's without imputation.",
					icon = "warning")
        } else {dat = dat2}
      }
    }
    if (method == 'Regression') {
      if (sum(complete.cases(dat))!=0) {
        formula0 = as.formula(paste('~ ',paste(names(dat),collapse=' + ')))
        f = aregImpute(formula0, data=dat)
        tmpres = f$imputed
        for (i in 1:length(tmpres)) {
          dat[rownames(tmpres[[i]]),names(tmpres)[i]]=tmpres[[i]][,1]
        }
      }
    }
    if (method == 'Nearest neighbor') {
      if (sum(complete.cases(dat))!=0 & (!("character" %in% vartype))) {
        myNNdat = dat[complete.cases(dat),]
        Missing_any = factor(apply(shadow,1,any),
                             levels=c(FALSE,TRUE),labels=c('Not','Missing'))
        for (i in which(Missing_any=='Missing')){
          usecol = which(!is.na(dat[i,]))
          if (length(usecol)!=0){
            NNdat = myNNdat
            a = rbind(dat[i,], NNdat)[,usecol]
            NNdat$distance = dist(a)[1:nrow(NNdat)]
            k5NNdat = NNdat[order(NNdat$distance,decreasing=FALSE),][1:min(5,nrow(NNdat)),]
            if (nrow(k5NNdat)>1 & n-length(usecol)>1) {
              dat[,1:n][i,-usecol] = apply(k5NNdat[,1:n][,-usecol],2,mean)
            } else {
              if (nrow(k5NNdat)==1) {
                dat[,1:n][i,-usecol] = k5NNdat[,1:n][1,-usecol]
              } else {
                dat[,1:n][i,-usecol] = mean(k5NNdat[,1:n][,-usecol])
              }
            }
          } else {
            for (j in 1:n) {
              dat[i,j] = median(dat[,j], na.rm=TRUE)
            }
          }
        }
      }
    }
    if (method == 'Mode') {
      if (any(c('integer','numeric') %in% vartype)) {
        gmessage("The variables you selected contain numeric data. But I don't want to compute mode for numeric variables.",
                 icon = "warning")
      } else {
        for (i in 1:n) {
          dat[shadow[,i],i] = names(sort(table(na.omit(dat[,i])),decreasing=TRUE))[1]
        }
      }
    }
  } else {
    missingpct = mean(is.na(dat))
    if (method == 'Below 10%') {
      if (vartype %in% c('integer','numeric','logical') &
        missingpct>0) {
        dat[shadow] = min(dat, na.rm=TRUE)*1.1-
          max(dat,na.rm=TRUE)/10
      }
      if (vartype=='character') {
        dat[shadow] = 'NAN'
      }
      if (vartype=='factor' & missingpct<1 &
        missingpct>0) {
        dat=factor(dat,levels = c('NAN',levels(factor(dat))))
        dat[shadow] = 'NAN'
      }
      if (vartype=='factor' & missingpct==0) {
        dat=factor(dat)
      }
    }
    if (method == 'Median') {
      dat = as.numeric(impute(dat, fun=median))
    }
    if (method == 'Mean') {
      dat = as.numeric(impute(dat, fun=mean))
    }
    if (method == 'Random value') {
      dat = as.numeric(impute(dat, fun='random'))
    }
    if (method == 'Mode') {
      if (any(c('integer','numeric') %in% vartype)) {
        gmessage("The variables you selected contain numeric data. But I don't want to compute mode for numeric variables.",
                 icon = "warning")
        dat = NULL
      } else {
        dat[shadow] = names(sort(table(na.omit(dat)),decreasing=TRUE))[1]
      }
    }
  }
  return(data.frame(dat,row_number=row_number_2011))
}
